File: 10KindsOfPeople/C#/Program.cs
using System;
using System.Collections.Generic;

namespace BinaryOrDecimalCSharp
{
    class Program
    {
        static void Main(string[] args)
        {
            var firstLine = Console.ReadLine().Split(" ");
            var d1 = int.Parse(firstLine[0]);
            var d2 = int.Parse(firstLine[1]);

            // setup our playing field
            var patternLines = Read(d1);
            var field = new int[d1, d2];
            for(int i = 0; i < d1; i++)
            {
                int y = 0;
                foreach(int value in ConvertStringToIntArray(patternLines[i]))
                {
                    field[i, y++] = value;
                }
            }
            
            // now field should be all set

            // create array to hold group data, and init group numbers to -1
            var disjoint = new int[d1*d2];
            for (int i = 0; i < disjoint.Length; i++)
                disjoint[i] = -1;
            
            // add each in the array to Union Find (two should be merged, if the value of the field are the same)
            for (int i = 0; i < d1; i++)
            {
                for(int j = 0; j < d2; j++)
                {
                    int ti, tj;

                    ti = i + 1;
                    tj = j;
                    if (InRange(d1, d2, ti, tj) && field[i, j] == field[ti, tj])
                    {
                        Join(disjoint, i * d2 + j, ti * d2 + tj);
                    }

                    ti = i;
                    tj = j + 1;
                    if (InRange(d1, d2, ti, tj) && field[i, j] == field[ti, tj])
                    {
                        Join(disjoint, i * d2 + j, ti * d2 + tj);
                    }

                }
            }

            // parse number of paths, and paths
            var numberOfPaths = int.Parse(Console.ReadLine());
            var pathLines = Read(numberOfPaths);

            // process paths
            foreach(string pathString in pathLines)
            {
                // check if the two points are in the same group, if no print which
                var arrPath = pathString.Split(" ");
                var (r1, c1) = (int.Parse(arrPath[0]) -1, int.Parse(arrPath[1]) - 1);   // start
                var (r2, c2) = (int.Parse(arrPath[2]) -1, int.Parse(arrPath[3]) - 1);   // end

                if (Find(disjoint, r1*d2+c1) == Find(disjoint, r2*d2+c2))
                {
                    if (field[r1, c1] == 1)
                    {
                        Console.Out.WriteLine("decimal");
                    }
                    else
                    {
                        Console.Out.WriteLine("binary");
                    }
                }
                else
                {
                    Console.Out.WriteLine("neither");
                }
            }


        }

        static bool InRange(int x, int y, int ti, int tj)
        {
            if (ti < 0)
            {
                return false;
            }
            if (tj < 0)
            {
                return false;
            }

            if (ti >= x)
            {
                return false;
            }
            if (tj >= y)
            {
                return false;
            }

            return true;
        }

        static int Find(int[] disjoint, int a)
        {
            // default case
            if(disjoint[a] == -1)
            {
                return a;
            }

            disjoint[a] = Find(disjoint, disjoint[a]);
            return disjoint[a];
        }

        static void Join(int[] disjoint, int a, int b)
        {
            a = Find(disjoint, a);
            b = Find(disjoint, b);

            if (a == b)
                return;

            disjoint[a] = b;

        }

        static List<string> Read(int n)
        {
            var list = new List<string>();
            for (int i = 0; i < n; i++)
                list.Add(Console.ReadLine());
            return list;
        }

        static IEnumerable<int> ConvertStringToIntArray(string s)
        {
            for (int i = 0; i < s.Length; i++)
                yield return int.Parse(s.Substring(i, 1));
        }

    }
}

File: 10KindsOfPeople/C/Program.c
#include <stdio.h>

int inRange(int x, int y, int ti, int tj)
{
    if (ti < 0)
    {
        return 0;
    }
    if (tj < 0)
    {
        return 0;
    }

    if (ti >= x)
    {
        return 0;
    }
    if (tj >= y)
    {
        return 0;
    }

    return 1;
}

int find(int disjoint[], int a)
{
    // default case
    if(disjoint[a] == -1)
    {
        return a;
    }

    disjoint[a] = find(disjoint, disjoint[a]);
    return disjoint[a];
}

void join(int disjoint[], int a, int b)
{
    a = find(disjoint, a);
    b = find(disjoint, b);

    if (a == b)
        return;

    disjoint[a] = b;
}

int main()
{
    int d1, d2;
    scanf("%d %d", &d1, &d2);

    int field[d1][d2];
    for(int i = 0; i < d1; i++)
    {
        char arr[d2];
        scanf("%s", &arr);
        for(int y = 0; y < d2; y++)
        {
            field[i][y] = arr[y] - '0';
        }
    }

    // now field should be all set

    // create array to hold group data, and init group numbers to -1
    int len = d1*d2;
    int disjoint[len];
    for (int i = 0; i < len; i++)
        disjoint[i] = -1;
    
    // add each in the array to Union Find (two should be merged, if the value of the field are the same)
    for (int i = 0; i < d1; i++)
    {
        for(int j = 0; j < d2; j++)
        {
            int ti, tj;

            ti = i + 1;
            tj = j;
            if (inRange(d1, d2, ti, tj) && field[i][j] == field[ti][tj])
            {
                join(disjoint, i * d2 + j, ti * d2 + tj);
            }

            ti = i;
            tj = j + 1;
            if (inRange(d1, d2, ti, tj) && field[i][j] == field[ti][tj])
            {
                join(disjoint, i * d2 + j, ti * d2 + tj);
            }

        }
    }

    // parse number of paths, and paths
    int numberOfPaths;
    scanf("%d", &numberOfPaths);
    
    int pathLines[numberOfPaths][4];
    for(int i = 0; i < numberOfPaths; i++)
    {
        scanf("%d %d %d %d", &pathLines[i][0], &pathLines[i][1], 
        &pathLines[i][2], &pathLines[i][3]);
    }

    // process paths
    for(int i = 0; i < numberOfPaths; i++)
    {
        int r1, c1, r2, c2;
        r1 = pathLines[i][0] - 1;
        c1 = pathLines[i][1] - 1;
        r2 = pathLines[i][2] - 1;
        c2 = pathLines[i][3] - 1;
        if (find(disjoint, r1*d2+c1) == find(disjoint, r2*d2+c2))
        {
            if (field[r1][c1] == 1)
            {
                printf("decimal\n");
            }
            else
            {
                printf("binary\n");
            }
        }
        else
        {
            printf("neither\n");
        }
    }
    
    return 0;

}
